dbms/src/Common/UniThreadPool.h

```diff
@@ -284,6 +287,75 @@ class ThreadFromGlobalPoolImpl : boost::noncopyable
     }
 };
 
+/// ThreadPoolWaitGroup is used to wait all the task launched here to finish
+/// ThreadPoolWaitGroup guarantee the exception safty of TheadPool.
+template <typename ThreadPool>
+class ThreadPoolWaitGroup
+{
+public:
+    explicit ThreadPoolWaitGroup(ThreadPool * thread_pool_)
+        : thread_pool(thread_pool_)
+    {}
+    ThreadPoolWaitGroup(const ThreadPoolWaitGroup &) = delete;
+    ~ThreadPoolWaitGroup()
+    {
+        try
+        {
+            wait();
+        }
+        catch (const Exception & exc)
+        {
+            LOG_ERROR(&Poco::Logger::get("ThreadPoolWaitGroup"), "Exception error code = {}, message = {}", exc.code(), exc.displayText());
+        }
+    }
+
+    void schedule(std::shared_ptr<std::packaged_task<void()>> task)
+    {
+        thread_pool->scheduleOrThrowOnError([task] { (*task)(); });
+        futures.emplace_back(task->get_future());
+    }
+
+    void wait()
+    {
+        if (consumed)
+            return;
+        consumed = true;
+
+        std::exception_ptr first_exception;
+        for (auto & future : futures)
+        {
+            // ensure all futures finished
+            try
+            {
+                future.get();
+            }
+            catch (...)
+            {
+                if (!first_exception)
+                    first_exception = std::current_exception();
+            }
+        }
+
+        if (first_exception)
+        {
+            try
+            {
+                std::rethrow_exception(first_exception);
+            }
+            catch (Exception & exc)
+            {
+                exc.addMessage(exc.getStackTrace().toString());
+                exc.rethrow();
+            }
+        }
+    }
+
+private:
+    std::vector<std::future<void>> futures;
+    ThreadPool * thread_pool;
```

Comment:
Better to be a `ThreadPool &` because we don't allow thread_pool to be not exist.